presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java

```diff
@@ -457,6 +516,79 @@ private ShardInfo createShardInfo(UUID shardUuid, OptionalInt bucketNumber, Path
         return shardDelta(shardUuid, Optional.of(shard));
     }
 
+    @VisibleForTesting
+    Collection<Slice> writeDeltaDeleteFile(long transactionId, OptionalInt bucketNumber, UUID oldShardUuid, Optional<UUID> oldDeltaDeleteShardUuid, BitSet rowsToDelete)
+    {
+        if (rowsToDelete.isEmpty()) {
+            return ImmutableList.of();
+        }
+        // blockToDelete is LongArrayBlock
+        List<Long> columnIds = new ArrayList<>();
+        List<Type> columnTypes = new ArrayList<>();
+        columnIds.add(0L);
+        columnTypes.add(BIGINT);
+
+        // todo one block can't hold one orcfile
+        // At this point, rowsToDelete couldn't be empty
+        boolean oldShardContainToDelete = false;
+        if (oldDeltaDeleteShardUuid.isPresent()) {
+            oldShardContainToDelete = mergeRowsToDelete(rowsToDelete, oldDeltaDeleteShardUuid.get());
+        }
+
+        if (rowsToDelete.cardinality() == getRowCount(oldShardUuid)) {
+            // Delete original file
+            return shardDeleteDelta(oldShardUuid, oldDeltaDeleteShardUuid, Optional.empty());
+        }
+        if (oldShardContainToDelete) {
+            return ImmutableList.of();
+        }
+
+        BlockBuilder blockBuilder = new LongArrayBlockBuilder(null, rowsToDelete.size());
+        for (int i = rowsToDelete.nextSetBit(0); i >= 0; i = rowsToDelete.nextSetBit(i + 1)) {
+            blockBuilder.writeLong(i);
+        }
+        Block blockToDelete = blockBuilder.build();
+        StoragePageSink pageSink = createStoragePageSink(transactionId, bucketNumber, columnIds, columnTypes, true);
+        pageSink.appendPages(ImmutableList.of(new Page(blockToDelete)));
+        List<ShardInfo> shardInfos = getFutureValue(pageSink.commit());
+        // Guaranteed that shardInfos only has one element since we only call commit one time
+        ShardInfo newDeltaDeleteShard = Iterables.getOnlyElement(shardInfos);
+        return shardDeleteDelta(oldShardUuid, oldDeltaDeleteShardUuid, Optional.of(newDeltaDeleteShard));
+    }
+
+    private boolean mergeRowsToDelete(BitSet rowsToDelete, UUID oldDeltaDeleteShardUuid)
+    {
+        Optional<BitSet> rowsDeleted = getRowsFromUuid(Optional.of(oldDeltaDeleteShardUuid));
+        if (rowsDeleted.isPresent()) {
+            boolean oldShardSizeBiggerThanToDelete = rowsDeleted.get().cardinality() >= rowsToDelete.cardinality();
+            int rowsDeletedOriginalSize = rowsDeleted.get().cardinality();
+            rowsToDelete.or(rowsDeleted.get());
+            return oldShardSizeBiggerThanToDelete && rowsToDelete.cardinality() == rowsDeletedOriginalSize;
```

Comment:
Why are we tracking this? Looks like you assume that if the old delta's cardinality is greater or equal to the new delta, there is no need to generate a new delta. But I think it's meaningless to compare their sizes. 

I think we should assert that the `rowsToDelete` and `rowsDeleted` are disjoint. You cannot delete a row twice.